File: src/Scene.rs
use Transformable::Transformable;

pub struct Scene {}

impl Scene {

    pub fn new () -> Scene {
        Scene {}
    }

    pub fn canvasConfiguration (self) -> String {
        String::from("
            const canvas = document.getElementById('target-canvas');
            canvas.width = canvas.clientWidth;
            canvas.height = canvas.clientHeight;
            const webgl = canvas.getContext('webgl');
            webgl.getExtension('OES_standard_derivatives');
            webgl.enable(webgl.DEPTH_TEST);
            webgl.depthFunc(webgl.LEQUAL);
            webgl.clear(webgl.DEPTH_BUFFER_BIT);
            webgl.clear(webgl.COLOR_BUFFER_BIT);
            webgl.viewport(0, 0, canvas.clientWidth, canvas.clientHeight);
        ")
    }

    pub fn getTransformations (self) -> String {
        String::from("
            const aspect = canvas.width / canvas.height;
            const a = 1 * Math.tan(45 * Math.PI / 360);
            const b = a * aspect;
            const h = b + b, i = a + a, j = 10 - 1;
            webgl.uniformMatrix4fv(webgl.getUniformLocation(shaderProgram, 'pMatrix'), false, [
                1 * 2 / h, 0, 0, 0,
                0, 1 * 2 / i, 0, 0,
                0, 0, -11 / j, -1,
                0, 0, -(10 * 1 * 2) / j, 0
            ]);
            webgl.uniformMatrix4fv(webgl.getUniformLocation(shaderProgram, 'uPMVMatrix'), false, [
                1, 0, 0, 0,
                0, 1, 0, 0,
                0, 0, 1, 0,
                0, 0, -5, 1
            ]);
        ")
    }

}

impl Transformable for Scene {

    fn translate (&mut self, position: Vec<f32>) {}

    fn rotate (&mut self, deg: f32, axis: Vec<f32>) {}

    fn scale (&mut self, scale: f32) {}

}

File: src/entity/Scene.js
const Entity = require('./Entity');
const Cubemap = require('./Cubemap.js');
const Shaders = require('../shader/Shaders.js');
const Lightmap = require('../structure/Lightmap')
const Cameras = require('../structure/Cameras')
const Render = require('../structure/Render.js');
const Events = require('../structure/Events');

/*
* @title Scene
*
* @description 
*
* @arg 
*
* @return Scene
*/
class Scene extends Entity {

    static getConfig () {
        return ({
            isUniqueInstance: true, 
            plural: 'scenes',
            singular: 'scene',
            defaults: {
                cameras: NativeTypes.infer([]),
                geometries: NativeTypes.infer([]),
                lights: NativeTypes.infer([]),
                cubemap: NativeTypes.infer(new Cubemap()),
            }
        });
    }

    getGeometries () {
        return this.geometries;
    }

    toString () {
        let out = `
            function viewer (container) {

                /*
                * Canvas Configuration
                */
                const canvas = document.createElement('canvas');
                canvas.style.width = "100%";
                canvas.style.height = "100%";
                canvas.width = container.clientWidth;
                canvas.height = container.clientHeight;
                container.appendChild(canvas);
                canvas.oncontextmenu = () => false;

                /*
                * WebGL Context Configuration
                */
                const webgl = canvas.getContext("webgl");
                webgl.viewport(0, 0, container.clientWidth, container.clientHeight);
                webgl.clearColor(0.0, 0.0, 0.0, 0.0);
                webgl.getExtension('OES_element_index_uint');
                webgl.getExtension('OES_standard_derivatives');
                webgl.getExtension('WEBGL_depth_texture');
                webgl.enable(webgl.DEPTH_TEST);
                webgl.depthFunc(webgl.LEQUAL);
                /*webgl.enable(webgl.BLEND);
                webgl.blendEquation(webgl.FUNC_ADD);
                webgl.blendFuncSeparate(webgl.SRC_ALPHA, webgl.ONE_MINUS_SRC_ALPHA, webgl.ONE, webgl.ONE_MINUS_SRC_ALPHA);*/
                //webgl.enable (webgl.CULL_FACE);

        `;

        const events = new Events();
        const lightmap = new Lightmap(this.lights);
        const cameras = new Cameras(this.cameras);
        const shaders = new Shaders({
            shouldRenderCubeMap: this.cubemap.shouldRenderCubeMap() && 
                                 this.cubemap.cubemap_reflection
        }, this.lights);

        if (this.cubemap) {
            out += this.cubemap.toString();
        }   

        for (const geometry of this.geometries) {
            out += geometry.toString();

            lightmap.addGeometry(geometry);
            shaders.addGeometry(geometry);
            events.addEvents(geometry.getEvents(), geometry);
        }

        for (const camera of this.cameras) {
            events.addEvents(camera.getEvents(), camera);
        }

        out += cameras.toString();
        out += lightmap.toString();
        out += shaders.toString();
        out += Render([ this.cubemap.generateRenderBlock(), shaders.generateRenderBlock() ]);
        out += events.toString();
        
        
        return out + '}';
    }

}

module.exports = Scene;